public/app/plugins/panel/datagrid/DataGridPanel.tsx

```diff
@@ -0,0 +1,485 @@
+          readonly: false,
+          displayData: value !== null && value !== undefined ? value.toString() : '',
+        };
+      default:
+        return {
+          kind: GridCellKind.Text,
+          data: value ? value : '',
+          allowOverlay: isDatagridEditEnabled()!,
+          readonly: false,
+          displayData: value !== null && value !== undefined ? value.toString() : '',
+        };
+    }
+  };
+
+  const onCellEdited = (cell: Item, newValue: EditableGridCell) => {
+    // If there is a gridSelection, then the only possible action for it was a multiple cell clear edit, which
+    // is handled by the onDeletePressed handler. So, we can return early and avoid performance issues if
+    // gridSelection has height 1 and width 1, meaning the user activated a single cell for edit.
+    if (
+      gridSelection.current?.range &&
+      gridSelection.current.range.height > 1 &&
+      gridSelection.current.range.width > 1
+    ) {
+      return;
+    }
+
+    const [col, row] = cell;
+    const field: Field = frame.fields[col];
+
+    if (!field) {
+      return;
+    }
+
+    const values = field.values.toArray();
+
+    values[row] = newValue.data;
+    field.values = new ArrayVector(values);
+
+    publishSnapshot(new MutableDataFrame(frame), id);
+  };
+
+  const onColumnInputBlur = (columnName: string) => {
+    const len = frame.length ?? 0;
+    const newFrame = new MutableDataFrame(frame);
+
+    const field: Field = {
+      name: columnName,
+      type: FieldType.string,
+      config: {},
+      values: new ArrayVector(new Array(len).fill('')),
+    };
+
+    newFrame.addField(field);
+
+    publishSnapshot(newFrame, id);
+  };
+
+  const addNewRow = () => {
+    const newFrame = new MutableDataFrame(frame);
+    newFrame.appendRow(new Array(newFrame.fields.length).fill(null));
+
+    publishSnapshot(newFrame, id);
+  };
+
+  const onColumnResize = (column: GridColumn, newSize: number, colIndex: number, newSizeWithGrow: number) => {
+    setColumns((prevColumns) => {
+      const newColumns = [...prevColumns];
+
+      newColumns[colIndex] = {
+        ...column,
+        width: newSize,
+      };
+
+      return newColumns;
+    });
+
+    setIsResizeInProgress(true);
+  };
+
+  const onColumnResizeEnd = (column: GridColumn, newSize: number, colIndex: number, newSizeWithGrow: number) => {
+    setIsResizeInProgress(false);
+  };
+
+  const closeContextMenu = () => {
+    setContextMenuData({ isContextMenuOpen: false });
+  };
+
+  const onDeletePressed = (selection: GridSelection) => {
+    if (selection.current && selection.current.range) {
+      publishSnapshot(clearCellsFromRangeSelection(frame, selection.current.range), id);
+      return true;
+    }
+
+    if (selection.rows) {
+      publishSnapshot(deleteRows(frame, selection.rows.toArray()), id);
+      return true;
+    }
+
+    return false;
+  };
+
+  const onCellContextMenu = (cell: Item, event: CellClickedEventArgs) => {
```

Comment:
Seems like onCellContextMenu/onHeaderContextMenu/onHeaderContextMenu has a bunch of shared code, just a nitpick, but maybe refactor into a single function and pass options?

Reply:
I agree that there's shared code, but creating a single function for this feels like complicating things a bit for just a couple of lines